### PR TITLE
Rework uvc gadget

### PR DESCRIPTION
**Pull request by DanielGuramulta**

### CODE REVIEW

File: sdk/src/target/local_device.cpp [hunk 1]
```diff
@@ -766,3 +766,74 @@ aditof::Status LocalDevice::applyCalibrationToFrame(uint16_t *frame,
...
+        return Status::OK;
+    }
+
+    return Status::INVALID_ARGUMENT;
+}
+
+aditof::Status LocalDevice::queueBuffer(unsigned int index) {
+    using namespace aditof;
+
+    struct v4l2_buffer buf;
```
Comment (dNechita) [inline sdk/src/target/local_device.cpp]: Shouldn't the buf be cleared (set all bits to 0)?

File: sdk/src/target/local_device.cpp [hunk 2]
```diff
@@ -766,3 +766,74 @@ aditof::Status LocalDevice::applyCalibrationToFrame(uint16_t *frame,
...
+    buf.m.planes = m_implData->planes;
+    buf.index = index;
+
+    if (xioctl(m_implData->fd, VIDIOC_QBUF, &buf) == -1) {
+        LOG(WARNING) << "VIDIOC_QBUF error "
+                     << "errno: " << errno << " error: " << strerror(errno);
+        return Status::GENERIC_ERROR;
+    }
+
+    // LOG(INFO) << "Queued v4l2 buffer from LocalDevice!";
```
Comment (dNechita) [inline sdk/src/target/local_device.cpp]: Uncomment or remove

File: apps/uvc-app/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+cmake_minimum_required(VERSION 2.8)
+project(uvc-app)
```
Comment (dNechita) [inline apps/uvc-app/CMakeLists.txt]: What is the reason for changing the uvc-gadget name to uvc-app?

File: apps/uvc-app/uvc-gadget.cpp [hunk 4]
```diff
@@ -2047,18 +1307,24 @@ static void uvc_events_process_control(
...
                 resp->length = len;
                 break;
 
             case UVC_GET_CUR:
                 USB_REQ_DEBUG("Received GET_CUR on %d\n", cs);
 
-                if(cs == 5) {
-					eeprom_read_buf(&edev, eeprom_read_addr, (unsigned char*)resp->data, eeprom_read_len);
+                if (cs == 5) {
+                    // eeprom_read_buf(&edev, eeprom_read_addr,
```
Comment (dNechita) [inline apps/uvc-app/uvc-gadget.cpp]: These commented lines can be removed, right?

File: apps/uvc-app/uvc-gadget.cpp [hunk 5]
```diff
@@ -2311,81 +1592,71 @@ static int uvc_events_process_data(struct uvc_device *dev, struct uvc_request_da
...
+                /* Store the firmware data from Host and flash it when enough is
+                 * available */
+
+                // printf("store firmware data\n");
+                static bool shouldRestartDevice = false;
+
+                if (DeviceStartedStreaming) {
+                    shouldRestartDevice = true;
+                    DeviceStartedStreaming = false;
+                    // device->stop();
```
Comment (dNechita) [inline apps/uvc-app/uvc-gadget.cpp]: ?

File: apps/uvc-app/uvc-gadget.cpp [hunk 6]
```diff
@@ -2311,81 +1592,71 @@ static int uvc_events_process_data(struct uvc_device *dev, struct uvc_request_da
...
-                    {
-                        memset(&firmware[flen], 0, 4096 - flen);
-                        if (v4l2_subdev_program(firmware) < 0) {
-                            printf("Error Programing the chip\n");
-                        } else {
-                            printf("Flashed firmware, flen %d\n", flen);
-                        }
-                        flen = 0;
+                    if (shouldRestartDevice) {
+                        // device->start();
```
Comment (dNechita) [inline apps/uvc-app/uvc-gadget.cpp]: ?

File: sdk/src/target/local_device.cpp (hunk 1 above)
Comment (DanielGuramulta) [reply sdk/src/target/local_device.cpp]: It should. I've added the clear now. Thanks

File: apps/uvc-app/CMakeLists.txt (hunk 3 above)
Comment (DanielGuramulta) [reply apps/uvc-app/CMakeLists.txt]: Named it wrong when I wrote the `CMakeLists.txt`. Should be fixed now.

File: sdk/src/local_device.h [hunk 7]
```diff
@@ -36,6 +36,13 @@ class LocalDevice : public DeviceInterface {
     virtual aditof::Status applyCalibrationToFrame(uint16_t *frame,
                                                    const std::string &mode);
 
+    aditof::Status getDeviceFileDescriptor(int &fileDescriptor);
```
Comment (dNechita) [inline sdk/src/local_device.h]: these methods do not fall in the same category (`public: // implements DeviceInterface`) as the rest of the methods. You could add another public modifier for these calls to make things more clear.

Comment (DanielGuramulta) [reply sdk/src/local_device.h]: I have added another `public` modifier for these methods. Thanks!

File: apps/uvc-app/uvc-gadget.cpp [hunk 8]
```diff
@@ -241,760 +236,136 @@ struct uvc_device {
...
-{
-    close (v4l2_subdev_fd);
-}
+    if (!udev->hasNewFrame) {
+        return 0;
+    }
 
-static int readAfeTemp(float* temperature) {
-    temp_sensor tdev;
+    auto status = static_cast<LocalDevice *>(device)->dequeueBuffer(
```
Comment (dNechita) [inline apps/uvc-app/uvc-gadget.cpp]: The safely downcast use dynamic_cast

Comment (DanielGuramulta) [reply apps/uvc-app/uvc-gadget.cpp]: I've changed the code to use `dynamic_cast`.

**Pull request merged**